File: 20151202/src/enshu_20151202/src/GCTest.cpp
#include <iostream>
#include <vector>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <pthread.h>
#include <sys/time.h>

using namespace std;

void *_ThreadStart(void *arg);
class Thread {
public:
  Thread(){};
  int Start(){
    Setup();
    int code = pthread_create(&ThreadId_, NULL, _ThreadStart, (void *)this);
    return code;
  }
  int Join(){
    int code = pthread_join(ThreadId_, NULL);
    return code;
  }
  virtual void Setup(){};
  virtual void Execute(){};
private:
  pthread_t ThreadId_;
  void * Arg_;
};
void *_ThreadStart(void *arg){
  ((Thread *)arg)->Execute();
}


///
///
///

bool do_delete = true;


typedef vector<int> Elem;
typedef vector<int>* ElemPtr;
typedef vector<ElemPtr>* ElemPtrVec;

class Mutator : public Thread {
  bool loop;
public:
  Mutator(){}
  void Setup (){
    loop = true;
  }
  void Execute(){
    while (loop) {
      ElemPtrVec v = ElemPtrVec(new vector<ElemPtr>());
      // mutator
      for (int k = 0; k < 2000; k++) {
	v->push_back(ElemPtr(new Elem(1000000)));
      }
      //
#if 0 // usually we need this
      if ( do_delete ) {
	for (vector<ElemPtr>::iterator i = v->begin();
	     i != v->end(); i++){
	       delete (*i);
	     }
	delete v;
      }
#endif
      usleep(0);
      cout << '\b' << "-";
    }
  }
    
  void end() {
    loop = false;
  }
};


class Sort : public Thread {
  struct timeval tv;
  struct timezone tz;
  float time;

  void print_arr(vector<int> a){
    for (int i = 0; i< a.size(); i+=1000) {
      cout << a[i];
    }
    gettimeofday(&tv, &tz);
    float elapsedtime =  tv.tv_sec*1000 + tv.tv_usec/1000.0;
    cout << " " << (elapsedtime-time) << " ms" << endl;;
    time = elapsedtime;
  }

  void scramble(vector<int> *a) {
    double f = 10.0 / (double) a->size();
    for (int i = a->size()-1; --i >= 0;) {
      (*a)[i] = (int)(i * f);
    }
    for (int i = a->size()-1; --i >= 0;) {
      int j = (int)(i * (float)random()/RAND_MAX);
      int t = (*a)[i];
      (*a)[i] = (*a)[j];
      (*a)[j] = t;
    }
  }

  void sort(vector<int> *a) {
    for (int i = a->size()-1; --i>=0; ) {
      bool swapped = false;
      for (int j = 0; j<i; j++) {
	if ((*a)[j] > (*a)[j+1]) {
	  int T = (*a)[j];
	  (*a)[j] = (*a)[j+1];
	  (*a)[j+1] = T;
	  swapped = true;
	}
      }
      if ( i % 10 == 0 ) print_arr(*a);
      if (!swapped)
	return;
    }
  }
  
public:
  vector<int> arr;
  Sort(){}
  void Setup (){
    arr.resize(50000);
  }
  void Execute(){
    cout << "** start Sort *****" << endl;
    scramble(&arr);
    print_arr(arr);
    sort(&arr);
    print_arr(arr);
    cout << "** done Sort ********" << endl;
  }
};

#include <gtest/gtest.h>
// Declare a test
TEST(TestSuite, testCase1)
{
  // dummy test
}

int main(int argc, char *argv[]){
  cout << argc << endl;
  if ( argc > 1) if (!strcmp(argv[1], "--no-delete")) {
    do_delete = false;
    cerr << "NO DELETE mode" << endl;
  }
  Sort *sort = new Sort();
  Mutator *mutator = new Mutator();
  sort->Start();
  mutator->Start();
  sort->Join();
  mutator->end();
  mutator->Join();

  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
  return 0;
}

File: 20151202/src/enshu_20151202/src/GCTest.java
import java.util.*;

class Mutator extends Thread {
  boolean loop = true;
  
  public void run() {
    while ( loop ) {
      Vector v = new Vector();
      // mutator
      for (int k = 0; k < 1000; k++) {
	Vector e = new Vector();
	for (int l = 0; l < 10000; l++) {
	  e.addElement(new Integer(l));
	}
	Thread.yield();
	v.addElement(e);
      }
      //
      System.out.print('\b'+"-");
      v = null;
    }
  }
    
  public void end() {
    loop = false;
  }
}

class Sort extends Thread {
  static int arr[];
  static long time;

  public static void print_arr(){
    for (int i = 0; i< arr.length; i+=300) {
      System.out.print(arr[i]);
    }
    System.out.println(" " + (System.currentTimeMillis()-time) + " ms");
    time = System.currentTimeMillis();
  }

  static void scramble() {
    int a[] = new int[21000];
    double f = 10.0 / (double) a.length;
    for (int i = a.length; --i >= 0;) {
      a[i] = (int)(i * f);
    }
    for (int i = a.length; --i >= 0;) {
      int j = (int)(i * Math.random());
      int t = a[i];
      a[i] = a[j];
      a[j] = t;
    }
    arr = a;
  }

  static void sort(int a[]) throws Exception {
    int y = 0;
    for (int i = a.length; --i>=0; ) {
      boolean swapped = false;
      for (int j = 0; j<i; j++) {
	if (a[j] > a[j+1]) {
	  int T = a[j];
	  a[j] = a[j+1];
	  a[j+1] = T;
	  swapped = true;
	}
      }
      Thread.yield();
      if ( i % 50 == 0 ) print_arr();
      if (!swapped)
	return;
    }
  }
  public void run(){
    System.out.println("** start Sort *****");
    scramble();
    print_arr();
    try {
      sort(arr);
    } catch(Exception e) {
    }
    print_arr();
    System.out.println("** done Sort ********");
  }
}
	
class GCTest {

  public static void main(String[] args) {
    Sort sort = new Sort();
    Mutator mutator = new Mutator();
    sort.start();
    mutator.start();
    try {
      sort.join();
      mutator.end();
      sort.print_arr();
    } catch (InterruptedException e) {
      System.out.println(e);
    }
  }
}

